symengine/tests/basic/test_functions.cpp

```diff
@@ -4081,3 +4085,147 @@ TEST_CASE("test_dummy", "[Dummy]")
     CHECK(neq(*xdummy1, *x1));
     CHECK(neq(*xdummy1, *x1->as_dummy()));
 }
+
+TEST_CASE("test_floor", "[Floor]")
+{
+    RCP<const Basic> x = symbol("x");
+    RCP<const Basic> r = floor(integer(1));
+    CHECK(eq(*r, *one));
+
+    r = floor(Complex::from_two_nums(*integer(2), *integer(1)));
+    CHECK(eq(*r, *Complex::from_two_nums(*integer(2), *integer(1))));
+
+    r = floor(Nan);
+    CHECK(eq(*r, *Nan));
+
+    r = floor(Inf);
+    CHECK(eq(*r, *Inf));
+
+    r = floor(NegInf);
+    CHECK(eq(*r, *NegInf));
+
+    r = floor(Rational::from_two_ints(3, 1));
+    CHECK(eq(*r, *integer(3)));
+
+    r = floor(Rational::from_two_ints(3, 2));
+    CHECK(eq(*r, *integer(1)));
+
+    r = floor(Rational::from_two_ints(-3, 2));
+    CHECK(eq(*r, *integer(-2)));
+
+    r = floor(real_double(2.65));
+    CHECK(eq(*r, *integer(2)));
+
+    r = floor(complex_double(std::complex<double>(2.86, 2.79)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *Complex::from_two_nums(*integer(2), *integer(2))));
+
+    r = floor(pi);
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *integer(3)));
+
+    r = floor(E);
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *integer(2)));
+
+    r = floor(floor(x));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *floor(x)));
+
+    r = floor(ceiling(x));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *ceiling(x)));
+
+    CHECK_THROWS_AS(floor(Eq(integer(2), integer(3))), SymEngineException);
+
+#ifdef HAVE_SYMENGINE_MPFR
+    mpfr_class a(100), b(100);
+    mpfr_set_flt(a.get_mpfr_t(), 10.65, MPFR_RNDN);
+    mpfr_set_flt(b.get_mpfr_t(), 11.47, MPFR_RNDN);
+    r = floor(real_mpfr(std::move(a)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *integer(10)));
+    r = floor(real_mpfr(std::move(b)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *integer(11)));
+#ifdef HAVE_SYMENGINE_MPC
+    mpc_class c(100);
+    mpc_set_fr_fr(c.get_mpc_t(), a.get_mpfr_t(), b.get_mpfr_t(), MPFR_RNDN);
+    r = floor(complex_mpc(std::move(c)));
+    std::cout << *r << std::endl;
+    CHECK(eq(*r, *Complex::from_two_nums(*integer(10), *integer(11))));
```

Comment:
@isuruf I couldn't make out why this case fails here, whereas, its counterpart in `ceiling` works flawlessly.